### PR TITLE
PMacc warp and atomic function refactor

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- introduce support for HIP
- change the way how `atomicAllInc` is implemented if no accelerator is
given
- use alpaka intrinsics and warp functions

### CODE REVIEW

File: include/pmacc/nvidia/atomic.hpp [hunk 1]
```diff
@@ -150,19 +152,21 @@ template<typename T, typename T_Acc, typename T_Hierarchy>
...
-    return atomicAllInc(alpaka::atomic::AtomicStdLibLock<16>(), ptr, ::alpaka::hierarchy::Grids());
-#endif
+    /* Dirty hack to call a alpaka accelerator based function.
+     * Member of  the fakeAcc will be uninitialized and should be not accessed.
+     *
+     * The id provider for particles is the only code where atomicAllInc without an accelerator is used.
+     * @todo remove the unsafe faked accelerator
+     */
+    pmacc::memory::Array<cupla::AccThreadSeq, 1> fakeAcc;
+    return atomicAllInc(fakeAcc[0], ptr, ::alpaka::hierarchy::Grids());
```
Comment (psychocoderHPC) [inline include/pmacc/nvidia/atomic.hpp:169]: @jkelling This is the new way how to handle the unknown accelerator in this function. You can remove your workaround with the pre-compiler if cascade in your SPEC branch after you rebased against this changes.

File: include/pmacc/nvidia/atomic.hpp [hunk 2]
```diff
@@ -23,12 +23,18 @@
...
+#include "pmacc/memory/Array.hpp"
+#if( BOOST_LANG_CUDA || BOOST_COMP_HIP)
 #   include "pmacc/nvidia/warp.hpp"
 #endif
 #include <boost/type_traits.hpp>
 #include <climits>
 
+#include <alpaka/intrinsic/Traits.hpp>
+#include <alpaka/warp/Traits.hpp>
+
```
Comment (PrometheusPi) [inline include/pmacc/nvidia/atomic.hpp:35]: Is the new line separation of the includes (line 32 and line 35) intended?

File: include/pmacc/nvidia/atomic.hpp [hunk 3]
```diff
@@ -150,19 +152,21 @@ template<typename T, typename T_Acc, typename T_Hierarchy>
...
 HDINLINE
 T atomicAllInc(T *ptr)
 {
-#ifdef __CUDA_ARCH__
-    return atomicAllInc(alpaka::atomic::AtomicUniformCudaHipBuiltIn(), ptr, ::alpaka::hierarchy::Grids());
-#else
-    // assume that we can use the standard library atomics if we are not on gpu
-    return atomicAllInc(alpaka::atomic::AtomicStdLibLock<16>(), ptr, ::alpaka::hierarchy::Grids());
-#endif
+    /* Dirty hack to call a alpaka accelerator based function.
```
Comment (PrometheusPi) [inline include/pmacc/nvidia/atomic.hpp:162]:
```suggestion
    /* Dirty hack to call an alpaka accelerator based function.
```

File: include/pmacc/nvidia/atomic.hpp [hunk 4]
```diff
@@ -150,19 +152,21 @@ template<typename T, typename T_Acc, typename T_Hierarchy>
...
+     * Member of the fakeAcc will be uninitialized and should be not accessed.
```
Comment (PrometheusPi) [inline include/pmacc/nvidia/atomic.hpp:163]:
```suggestion
     * Members of the fakeAcc will be uninitialized and should not be accessed.
```

File: include/pmacc/nvidia/atomic.hpp [hunk 5]
```diff
@@ -150,19 +152,21 @@ template<typename T, typename T_Acc, typename T_Hierarchy>
...
+     *
+     * The id provider for particles is the only code where atomicAllInc without an accelerator is used.
```
Comment (PrometheusPi) [inline include/pmacc/nvidia/atomic.hpp:165]:
```suggestion
     * The id provider for particles is the only code where atomicAllInc is used without an accelerator.
```

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
From my side, there a just some remarks on the comment text and code structure. For code validation, I would rely on @sbastrakov expertise.

File: include/pmacc/nvidia/atomic.hpp [hunk 6]
```diff
@@ -23,12 +23,18 @@
 
 
 #include "pmacc/types.hpp"
-#if( PMACC_CUDA_ENABLED == 1 )
+#include "pmacc/memory/Array.hpp"
+#if( BOOST_LANG_CUDA || BOOST_COMP_HIP)
 #   include "pmacc/nvidia/warp.hpp"
 #endif
```
Comment (sbastrakov) [inline include/pmacc/nvidia/atomic.hpp:29]: I am unaware of reasoning, so perhaps a stupid question. Ins't it better to put such guards inside this file and include it always?

File: include/pmacc/nvidia/atomic.hpp [hunk 7]
```diff
@@ -132,7 +134,7 @@ namespace nvidia
         template<typename T_Type>
         struct AtomicAllInc<T_Type, true>: public AtomicAllIncKepler<T_Type>
         {};
-#endif /* PMACC_CUDA_ARCH >= 300 */
+#endif /* CUPLA_DEVICE_COMPILE */
```
Comment (sbastrakov) [inline include/pmacc/nvidia/atomic.hpp:137]: Perhaps add  `== 1` in the end of the comment to match the condition at start?

File: include/pmacc/nvidia/atomic.hpp (hunk 2 above)
Comment (sbastrakov) [reply include/pmacc/nvidia/atomic.hpp:35]: I think it's now a little weird that standard library includes are separated by alpaka ones.

File: include/pmacc/nvidia/atomic.hpp (hunk 4 above)
Comment (sbastrakov) [reply include/pmacc/nvidia/atomic.hpp:163]: Or even `must not` be accessed

**Review by sbastrakov — COMMENTED** [review]
Nice, some comments inline. I also agree with @PrometheusPi 's wording suggestions.

**Comment by psychocoderHPC** [comment]
I integrated the suggestions

**Review by sbastrakov — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
can be merge

**Comment by sbastrakov** [comment]
I am fine with it, need @PrometheusPi to approve, otherwise the button is not clickable for me.

**Pull request merged**